mirrord-agent/src/dns.rs

```diff
@@ -1,87 +1,76 @@
-use std::path::PathBuf;
+use std::{
+    fs::{self, File},
+    path::{Path, PathBuf},
+};
 
 use mirrord_protocol::{
-    AddrInfoHint, AddrInfoInternal, GetAddrInfoRequest, RemoteResult, ResponseError,
+    dns::{DnsLookup, GetAddrInfoRequest, GetAddrInfoResponse},
+    RemoteResult,
 };
 use tokio::sync::{mpsc::Receiver, oneshot::Sender};
-use tracing::{error, trace};
+use tracing::{debug, error};
+use trust_dns_resolver::{system_conf::parse_resolv_conf, AsyncResolver, Hosts};
 
 use crate::{error::AgentError, runtime::set_namespace};
 
 #[derive(Debug)]
 pub struct DnsRequest {
     request: GetAddrInfoRequest,
-    tx: Sender<RemoteResult<Vec<AddrInfoInternal>>>,
+    tx: Sender<GetAddrInfoResponse>,
 }
 
 impl DnsRequest {
-    pub fn new(
-        request: GetAddrInfoRequest,
-        tx: Sender<RemoteResult<Vec<AddrInfoInternal>>>,
-    ) -> Self {
+    pub fn new(request: GetAddrInfoRequest, tx: Sender<GetAddrInfoResponse>) -> Self {
         Self { request, tx }
     }
 }
 
-trait AddrInfoHintExt {
-    fn into_lookup(self) -> dns_lookup::AddrInfoHints;
-}
+#[tracing::instrument(level = "debug")]
+async fn dns_lookup(root_path: &Path, host: String) -> RemoteResult<DnsLookup> {
+    let resolv_conf_path = root_path.join("etc").join("resolv.conf");
```

Comment:
I suggest to consider creating a resolver once and re-using it for performance, it's a bit of a nitpick but might have value down the road.
While writing this comment I thought about the situation where the files might change, so this needs to be hash-based, so maybe just a comment for later on improvement.

Reply:
Added a TODO there. Agree that we should be caching this, hopefully it doesn't hit us too hard for now.

Reply:
Probably not but it's good to have speed ups at hand in case we need :)